File: src/main/java/com/solarexsoft/rxjavainaction/DelayOperations.java
package com.solarexsoft.rxjavainaction;

import io.reactivex.Observable;
import io.reactivex.ObservableEmitter;
import io.reactivex.ObservableOnSubscribe;

/**
 * Created by Solarex on 2020/4/22/5:03 PM
 * Desc:
 */
public class DelayOperations {
    public static void main(String[] args) {
        Observable<Integer> observable1 = Observable.create(new ObservableOnSubscribe<Integer>() {
            @Override
            public void subscribe(ObservableEmitter<Integer> emitter) throws Exception {
                emitter.onNext(1);
                emitter.onNext(2);
                emitter.onNext(3);
                emitter.onNext(4);
                emitter.onComplete();
            }
        });

        Observable<String> observable2 = Observable.create(new ObservableOnSubscribe<String>() {
            @Override
            public void subscribe(ObservableEmitter<String> emitter) throws Exception {
                emitter.onNext("A");
                emitter.onNext("B");
                emitter.onNext("C");
                emitter.onComplete();
            }
        });
    }
}

File: src/main/kotlin/com/solarexsoft/rxjavainaction/DelayOperations.kt
package com.solarexsoft.rxjavainaction

import io.reactivex.Observable
import io.reactivex.rxkotlin.zipWith
import java.util.*
import java.util.concurrent.TimeUnit

/**
 * Created by Solarex on 2020/4/22/4:45 PM
 * Desc:
 */

fun delayOperations() {
    println("start ${Date()}")
    Observable.create<String> { s ->
        println("subscribing ${Date()}")
        s.onNext("a")
        s.onNext("b")
        s.onNext("c")
    }.doOnNext {
        println(">$it ${Date()}")
    }.delay(1000, TimeUnit.MILLISECONDS)
        .zipWith(
        Observable.range(1, 7).delay(2000, TimeUnit.MILLISECONDS).doOnNext { println(">$it ${Date()}") }.doOnComplete { println("done") }
    ).subscribe { println(it) }
}


fun main() {
    delayOperations()
}